mono/metadata/threads.c

```diff
@@ -1640,57 +1640,90 @@ ves_icall_System_Threading_Thread_GetName_internal (MonoInternalThread *this_obj
 	return str;
 }
 
-void 
-mono_thread_set_name_internal (MonoInternalThread *this_obj, MonoString *name, gboolean permanent, gboolean reset, MonoError *error)
+void
+mono_thread_set_name_internal (MonoInternalThread *this_obj, size_t length, const char* utf8, const gunichar2 *utf16, gboolean permanent, gboolean reset, MonoError *error)
 {
+	error_init (error);
+
+	// If length is set, then utf8 or utf16 or both must be set.
+	g_assert (!length || utf8 || utf16);
+
+	// Convert the utf8 or duplicate the utf16 outside the lock.
+	gunichar2* dup16 = NULL;
+	gunichar2* old_name = NULL;
+	char* dup8 = NULL;
+	GError *gerror = NULL;
 	MonoNativeThreadId tid = 0;
 
-	LOCK_THREAD (this_obj);
+	if (length > 0) {
+		if (utf16) {
+			dup16 = (gunichar2*)g_memdup (utf16, length * sizeof (gunichar2));
+		} else {
+			dup16 = g_utf8_to_utf16 (utf8, length, NULL, NULL, &gerror);
+			if (gerror) {
+				mono_error_set_execution_engine (error, "String conversion error: %s", gerror->message);
+				goto exit;
+			}
+		}
+	}
 
-	error_init (error);
+	LOCK_THREAD (this_obj);
 
 	if (reset) {
 		this_obj->flags &= ~MONO_THREAD_FLAG_NAME_SET;
 	} else if (this_obj->flags & MONO_THREAD_FLAG_NAME_SET) {
 		UNLOCK_THREAD (this_obj);
-		
 		mono_error_set_invalid_operation (error, "Thread.Name can only be set once.");
-		return;
-	}
-	if (this_obj->name) {
-		g_free (this_obj->name);
-		this_obj->name_len = 0;
+		goto exit;
 	}
-	if (name) {
-		this_obj->name = g_memdup (mono_string_chars (name), mono_string_length (name) * sizeof (gunichar2));
-		this_obj->name_len = mono_string_length (name);
 
+	old_name = this_obj->name;
+	this_obj->name = NULL;
+	this_obj->name_len = 0;
+
+	if (length > 0) {
+		this_obj->name = dup16;
+		dup16 = NULL;
+		this_obj->name_len = length;
 		if (permanent)
 			this_obj->flags |= MONO_THREAD_FLAG_NAME_SET;
 	}
-	else
-		this_obj->name = NULL;
 
 	if (!(this_obj->state & ThreadState_Stopped))
 		tid = thread_get_tid (this_obj);
 
 	UNLOCK_THREAD (this_obj);
 
 	if (this_obj->name && tid) {
-		char *tname = mono_string_to_utf8_checked (name, error);
-		return_if_nok (error);
-		MONO_PROFILER_RAISE (thread_name, ((uintptr_t)tid, tname));
-		mono_native_thread_set_name (tid, tname);
-		mono_free (tname);
+		const char *profiler_name = NULL;
+		if (utf8)
+			profiler_name = utf8;
+		else {
+			dup8 = mono_utf16_to_utf8 (utf16, length, error);
+			goto_if_nok (error, exit);
+			profiler_name = dup8;
+		}
+		MONO_PROFILER_RAISE (thread_name, ((uintptr_t)tid, profiler_name));
+		mono_native_thread_set_name (tid, profiler_name);
 	}
+exit:
+	g_free (old_name);
+	g_free (dup8);
+	g_free (dup16);
+	g_clear_error (&gerror);
 }
 
 void 
-ves_icall_System_Threading_Thread_SetName_internal (MonoInternalThread *this_obj, MonoString *name)
+ves_icall_System_Threading_Thread_SetName_internal (MonoInternalThreadHandle thread, MonoStringHandle name_handle, MonoError *error)
```

Comment:
Convert to UTF8 and use `fixed` in managed so you can simply call `mono_thread_set_name_internal` and so you can also simplify `mono_thread_set_name_internal` to only accept a utf8 string.

Reply:
My understanding is that we always want utf16 and sometimes also utf8. That is how the code was. We could convert preemptively in managed to utf8, fixed, like you say, and then use one or both. Or can/should we change what we remember in the internal thread as well? That makes sense I think. Note that on Windows 10 there is a new thread name feature that only accepts unicode. I added it. Anyway, ignore Windows for now I'll have to try this. I also wonder if we can't hold a long term gchandle and avoid the copy as well?

Reply:
Fair enough for needing UTF16 more often that UTF8, so let's only do the `fixed` work in managed to avoid introducing the `mono_unwrapped_string_*` concept.